### ISSUE TITLE
WS90 (Wittboy) + GW2000A: datetime error

### ISSUE THREAD

**Issue by alexdelprete** [bachya/ecowitt2mqtt#120]
**ecowitt2mqtt version**
2022.04.1

**Describe the bug**
Hi,

I have 2 ecowitt stations (WS80 with GW1000A gw and the new WS90 (Wittboy) with the GW2000A gateway) with a lot of extra sensors (T&H, soil, lightning, etc.). I'm trying to use ecowitt2mqtt with the Wittboy+GW2000A, and I'm using the following docker compose file to run ecowitt2mqtt:

```
version: "3"

services:

  ecowitt2mqtt:
    image: bachya/ecowitt2mqtt:latest
    hostname: ecowitt2mqtt
    container_name: ecowitt2mqtt
    restart: unless-stopped
    environment:
      PORT: "8080"
      INPUT_UNIT_SYSTEM: "imperial"
      OUTPUT_UNIT_SYSTEM: "metric"
      LOG_LEVEL: "DEBUG"
      MQTT_BROKER: "mqtt.axel.dom"
      MQTT_PORT: "1883"
      MQTT_USERNAME: "mqtt_pass"
      MQTT_PASSWORD: "mqtt_user"
      MQTT_TOPIC: "ecowitt2mqtt/GW2000A"
      ENDPOINT: "/data/report"
      # HASS_DISCOVERY: "False"
      # HASS_DISCOVERY_PREFIX: homeassistant
    ports:
      - "8880:8080"
```

When it receives data from the GW2000A, I see the following error in the log:

```
INFO:aiohttp.access:10.1.10.160 [20/Apr/2022:19:13:46 +0000] "POST /data/report HTTP/1.1" 500 244 "-" "-"
DEBUG:ecowitt2mqtt:Received data from Ecowitt device: {'PASSKEY': 'CYYYXXXXX057B', 'stationtype': 'GW2000A_V2.1.4', 'runtime': '436796', 'dateutc': '2022-04-20 19:14:47', 'tempinf': '72.9', 'humidityin': '56', 'baromrelin': '29.870', 'baromabsin': '29.509', 'tempf': '59.7', 'humidity': '65', 'winddir': '327', 'windspeedmph': '2.24', 'windgustmph': '3.80', 'maxdailygust': '17.45', 'solarradiation': '0.00', 'uv': '0', 'rainratein': '0.000', 'eventrainin': '0.000', 'hourlyrainin': '0.000', 'dailyrainin': '0.000', 'weeklyrainin': '0.000', 'monthlyrainin': '0.736', 'yearlyrainin': '3.909', 'rrain_piezo': '0.000', 'erain_piezo': '0.063', 'hrain_piezo': '0.000', 'drain_piezo': '0.075', 'wrain_piezo': '0.075', 'mrain_piezo': '0.941', 'yrain_piezo': '4.114', 'ws90cap_volt': '5.2', 'ws90_ver': '115', 'temp1f': '71.2', 'humidity1': '61', 'temp2f': '71.2', 'humidity2': '58', 'temp3f': '70.5', 'humidity3': '61', 'temp4f': '73.0', 'humidity4': '58', 'temp5f': '70.7', 'humidity5': '69', 'temp6f': '72.7', 'humidity6': '58', 'temp7f': '67.1', 'humidity7': '54', 'temp8f': '68.0', 'humidity8': '56', 'soilmoisture1': '53', 'soilmoisture2': '57', 'soilmoisture3': '59', 'soilmoisture4': '49', 'soilmoisture5': '52', 'pm25_ch1': '21.0', 'pm25_avg_24h_ch1': '16.3', 'tf_co2': '62.2', 'humi_co2': '61', 'pm25_co2': '4.9', 'pm25_24h_co2': '7.5', 'pm10_co2': '6.1', 'pm10_24h_co2': '7.8', 'co2': '455', 'co2_24h': '473', 'lightning_num': '13', 'lightning': '1', 'lightning_time': '1650475037', 'wh80batt': '3.28', 'batt1': '0', 'batt2': '0', 'batt3': '0', 'batt4': '0', 'batt5': '0', 'batt6': '0', 'batt7': '0', 'batt8': '0', 'soilbatt1': '1.4', 'soilbatt2': '1.3', 'soilbatt3': '1.3', 'soilbatt4': '1.3', 'soilbatt5': '1.3', 'pm25batt1': '3', 'wh57batt': '3', 'co2_batt': '6', 'wh90batt': '3.22', 'freq': '868M', 'model': 'GW2000A'}
DEBUG:ecowitt2mqtt:Wind Chill Temperature is only defined for temperatures at or below 50 F and wind speeds above 3 mph. (temperature: 59.7, wind speed: 2.24)
DEBUG:ecowitt2mqtt:Publishing entire device payload to single topic: ecowitt2mqtt/GW2000A
DEBUG:ecowitt2mqtt:Sending CONNECT (u1, p1, wr0, wq0, wf0, c1, k60) client_id=b''
DEBUG:ecowitt2mqtt:Received CONNACK (0, 0)
DEBUG:ecowitt2mqtt:Sending DISCONNECT
ERROR:aiohttp.server:Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
    resp = await request_handler(request)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/usr/src/ecowitt2mqtt/mqtt.py", line 116, in async_publish_payload
    await _async_publish_to_topic(client, data, topic)
  File "/usr/src/ecowitt2mqtt/mqtt.py", line 73, in _async_publish_to_topic
    await client.publish(topic, _generate_payload(data))
  File "/usr/src/ecowitt2mqtt/mqtt.py", line 19, in _generate_payload
    return json.dumps(data).encode("utf-8")
  File "/usr/local/lib/python3.9/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/local/lib/python3.9/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/lib/python3.9/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/local/lib/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type datetime is not JSON serializable
INFO:aiohttp.access:10.1.10.160 [20/Apr/2022:19:14:47 +0000] "POST /data/report HTTP/1.1" 500 244 "-" "-"
```

I used INPUT_UNIT_SYSTEM: "imperial" because I noticed from the raw data that the gw always sends imperial. Might the problem be the date/time format?

Thanks for any help on this.

Alessandro

**Comment by bachya** [comment]
I'm not entirely sure where the issue is coming from. Your 2nd log line shows the raw data received from your device; when I run a snippet of code equivalent to the code that currently throws the `TypeError`:

```python
import json

data = {
    "PASSKEY": "CYYYXXXXX057B",
    "stationtype": "GW2000A_V2.1.4",
    "runtime": "436796",
    "dateutc": "2022-04-20 19:14:47",
    "tempinf": "72.9",
    "humidityin": "56",
    "baromrelin": "29.870",
    "baromabsin": "29.509",
    "tempf": "59.7",
    "humidity": "65",
    "winddir": "327",
    "windspeedmph": "2.24",
    "windgustmph": "3.80",
    "maxdailygust": "17.45",
    "solarradiation": "0.00",
    "uv": "0",
    "rainratein": "0.000",
    "eventrainin": "0.000",
    "hourlyrainin": "0.000",
    "dailyrainin": "0.000",
    "weeklyrainin": "0.000",
    "monthlyrainin": "0.736",
    "yearlyrainin": "3.909",
    "rrain_piezo": "0.000",
    "erain_piezo": "0.063",
    "hrain_piezo": "0.000",
    "drain_piezo": "0.075",
    "wrain_piezo": "0.075",
    "mrain_piezo": "0.941",
    "yrain_piezo": "4.114",
    "ws90cap_volt": "5.2",
    "ws90_ver": "115",
    "temp1f": "71.2",
    "humidity1": "61",
    "temp2f": "71.2",
    "humidity2": "58",
    "temp3f": "70.5",
    "humidity3": "61",
    "temp4f": "73.0",
    "humidity4": "58",
    "temp5f": "70.7",
    "humidity5": "69",
    "temp6f": "72.7",
    "humidity6": "58",
    "temp7f": "67.1",
    "humidity7": "54",
    "temp8f": "68.0",
    "humidity8": "56",
    "soilmoisture1": "53",
    "soilmoisture2": "57",
    "soilmoisture3": "59",
    "soilmoisture4": "49",
    "soilmoisture5": "52",
    "pm25_ch1": "21.0",
    "pm25_avg_24h_ch1": "16.3",
    "tf_co2": "62.2",
    "humi_co2": "61",
    "pm25_co2": "4.9",
    "pm25_24h_co2": "7.5",
    "pm10_co2": "6.1",
    "pm10_24h_co2": "7.8",
    "co2": "455",
    "co2_24h": "473",
    "lightning_num": "13",
    "lightning": "1",
    "lightning_time": "1650475037",
    "wh80batt": "3.28",
    "batt1": "0",
    "batt2": "0",
    "batt3": "0",
    "batt4": "0",
    "batt5": "0",
    "batt6": "0",
    "batt7": "0",
    "batt8": "0",
    "soilbatt1": "1.4",
    "soilbatt2": "1.3",
    "soilbatt3": "1.3",
    "soilbatt4": "1.3",
    "soilbatt5": "1.3",
    "pm25batt1": "3",
    "wh57batt": "3",
    "co2_batt": "6",
    "wh90batt": "3.22",
    "freq": "868M",
    "model": "GW2000A",
}

print(json.dumps(data).encode("utf-8"))
```

...everything works as expected:

```python
b'{"PASSKEY": "CYYYXXXXX057B", "stationtype": "GW2000A_V2.1.4", "runtime": "436796", "dateutc": "2022-04-20 19:14:47", "tempinf": "72.9", "humidityin": "56", "baromrelin": "29.870", "baromabsin": "29.509", "tempf": "59.7", "humidity": "65", "winddir": "327", "windspeedmph": "2.24", "windgustmph": "3.80", "maxdailygust": "17.45", "solarradiation": "0.00", "uv": "0", "rainratein": "0.000", "eventrainin": "0.000", "hourlyrainin": "0.000", "dailyrainin": "0.000", "weeklyrainin": "0.000", "monthlyrainin": "0.736", "yearlyrainin": "3.909", "rrain_piezo": "0.000", "erain_piezo": "0.063", "hrain_piezo": "0.000", "drain_piezo": "0.075", "wrain_piezo": "0.075", "mrain_piezo": "0.941", "yrain_piezo": "4.114", "ws90cap_volt": "5.2", "ws90_ver": "115", "temp1f": "71.2", "humidity1": "61", "temp2f": "71.2", "humidity2": "58", "temp3f": "70.5", "humidity3": "61", "temp4f": "73.0", "humidity4": "58", "temp5f": "70.7", "humidity5": "69", "temp6f": "72.7", "humidity6": "58", "temp7f": "67.1", "humidity7": "54", "temp8f": "68.0", "humidity8": "56", "soilmoisture1": "53", "soilmoisture2": "57", "soilmoisture3": "59", "soilmoisture4": "49", "soilmoisture5": "52", "pm25_ch1": "21.0", "pm25_avg_24h_ch1": "16.3", "tf_co2": "62.2", "humi_co2": "61", "pm25_co2": "4.9", "pm25_24h_co2": "7.5", "pm10_co2": "6.1", "pm10_24h_co2": "7.8", "co2": "455", "co2_24h": "473", "lightning_num": "13", "lightning": "1", "lightning_time": "1650475037", "wh80batt": "3.28", "batt1": "0", "batt2": "0", "batt3": "0", "batt4": "0", "batt5": "0", "batt6": "0", "batt7": "0", "batt8": "0", "soilbatt1": "1.4", "soilbatt2": "1.3", "soilbatt3": "1.3", "soilbatt4": "1.3", "soilbatt5": "1.3", "pm25batt1": "3", "wh57batt": "3", "co2_batt": "6", "wh90batt": "3.22", "freq": "868M", "model": "GW2000A"}'
```

What's especially odd: that data contains only string values – there isn't a `datetime` object anywhere.

I'm curious if the data still gets published to the `ecowitt2mqtt/GW2000A` topic in your MQTT broker?

**Comment by alexdelprete** [comment]
> I'm curious if the data still gets published to the `ecowitt2mqtt/GW2000A` topic in your MQTT broker?

No, I already checked that when I opened the issue. Looks like it stops at the error, it's not a warning.

Please let me know what else I can do to help debugging this.

**Comment by bachya** [comment]
Okay, I see what's happening: the lightning strike epoch (which comes in as an `int`) gets converted to a `datetime` before the publish chokes. I'll address.

**Comment by alexdelprete** [comment]
Glad you found it. So I'm the only one using the WH57? :)

**Comment by bachya** [comment]
@alexdelprete Alright, I've confirmed this fixes your issue. The `dev` Docker image is building now.

**Comment by alexdelprete** [comment]
Ok, I'll test it and let you know. Thanks.

**Comment by alexdelprete** [comment]
It looked like it was working fine:

```json
INFO:ecowitt2mqtt:Published to ecowitt2mqtt/GW2000A: {'runtime': 1050395, 'tempin': 23.6, 'humidityin': 52, 'baromrel': 1017.813, 'baromabs': 1005.589, 'temp': 14.2, 'humidity': 69, 'winddir': 351, 'windspeed': 4.7, 'windgust': 6.5, 'maxdailygust': 16.78, 'solarradiation': 0.0, 'uv': 0, 'rainrate': 0.0, 'eventrain': 0.0, 'hourlyrain': 0.0, 'dailyrain': 0.0, 'weeklyrain': 0.0, 'monthlyrain': 36.8, 'yearlyrain': 117.4, 'rrain_piezo': 0.0, 'erain_piezo': 0.0, 'hrain_piezo': 0.0, 'drain_piezo': 0.4, 'wrain_piezo': 2.0, 'mrain_piezo': 76.2, 'yrain_piezo': 156.8, 'ws90cap_volt': 5.1, 'ws90_ver': 115, 'temp1': 22.3, 'humidity1': 61, 'temp2': 22.9, 'humidity2': 57, 'temp3': 21.9, 'humidity3': 65, 'temp4': 23.7, 'humidity4': 55, 'temp5': 21.5, 'humidity5': 66, 'temp6': 23.5, 'humidity6': 55, 'temp7': 20.6, 'humidity7': 48, 'temp8': 20.2, 'humidity8': 52, 'soilmoisture1': 53, 'soilmoisture2': 58, 'soilmoisture3': 60, 'soilmoisture4': 47, 'soilmoisture5': 47, 'pm25_ch1': 0.0, 'pm25_avg_24h_ch1': 0.0, 'tf_co2': 62.6, 'humi_co2': 58, 'pm25_co2': 4.9, 'pm25_24h_co2': 4.8, 'pm10_co2': 5.3, 'pm10_24h_co2': 5.2, 'co2': 497, 'co2_24h': 483, 'lightning_num': 98.2, 'lightning': 0.6, 'lightning_time': datetime.datetime(2022, 4, 27, 19, 1, 23, tzinfo=datetime.timezone.utc), 'wh80batt': 3.28, 'batt1': 'OFF', 'batt2': 'OFF', 'batt3': 'OFF', 'batt4': 'OFF', 'batt5': 'OFF', 'batt6': 'OFF', 'batt7': 'OFF', 'batt8': 'OFF', 'soilbatt1': 1.3, 'soilbatt2': 1.3, 'soilbatt3': 1.2, 'soilbatt4': 1.3, 'soilbatt5': 1.3, 'pm25batt1': 'ON', 'wh57batt': 'ON', 'co2_batt': 'ON', 'wh90batt': 3.22, 'dewpoint': 8.6, 'feelslike': 14.2, 'heatindex': 13.5, 'solarradiation_lux':0.0, 'solarradiation_perceived': 0.0, 'windchill': None}
```

![image](https://user-images.githubusercontent.com/7027842/165636086-cdbffaa8-84f4-414b-ae00-a480d8399f34.png)

Then I enabled HA discovery and got an error, and it stops sending data to MQTT. If I disable HA discovery, I can see updates in the MQTT topic again.

```json
DEBUG:ecowitt2mqtt:Using arguments: Namespace(log_level='DEBUG', mqtt_broker='mqtt.axel.dom', mqtt_port=1883, mqtt_username='xxxxxx', mqtt_password='xxxxx', mqtt_topic='ecowitt2mqtt/GW2000A', hass_discovery=True, hass_discovery_prefix='homeassistant', endpoint='/data/report', port=8080, raw_data=None, input_unit_system='imperial', output_unit_system='metric')
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:ecowitt2mqtt:Received data from Ecowitt device: {'PASSKEY': 'XXXXXXX', 'stationtype': 'GW2000A_V2.1.4', 'runtime': '1050700', 'dateutc': '2022-04-27 21:46:16', 'tempinf': '74.5', 'humidityin': '52', 'baromrelin': '30.056', 'baromabsin': '29.695', 'tempf': '57.6', 'humidity': '69', 'winddir': '353', 'windspeedmph': '4.92', 'windgustmph': '7.16', 'maxdailygust': '16.78', 'solarradiation': '0.00', 'uv': '0', 'rainratein': '0.000', 'eventrainin': '0.000', 'hourlyrainin': '0.000', 'dailyrainin': '0.000', 'weeklyrainin': '0.000', 'monthlyrainin': '1.449', 'yearlyrainin': '4.622', 'rrain_piezo': '0.000', 'erain_piezo': '0.000', 'hrain_piezo': '0.000', 'drain_piezo': '0.016', 'wrain_piezo': '0.079', 'mrain_piezo': '3.000', 'yrain_piezo': '6.173', 'ws90cap_volt': '5.1', 'ws90_ver': '115', 'temp1f': '72.1', 'humidity1': '61', 'temp2f': '73.0', 'humidity2': '57', 'temp3f': '71.4', 'humidity3': '65', 'temp4f': '74.7', 'humidity4': '55', 'temp5f': '70.7', 'humidity5': '65', 'temp6f': '74.3', 'humidity6': '55', 'temp7f': '68.9', 'humidity7': '48', 'temp8f': '68.2', 'humidity8': '52', 'soilmoisture1': '53', 'soilmoisture2': '58', 'soilmoisture3': '60', 'soilmoisture4': '47', 'soilmoisture5': '47', 'pm25_ch1': '0.0', 'pm25_avg_24h_ch1': '0.0', 'tf_co2': '62.4', 'humi_co2': '59', 'pm25_co2': '3.9', 'pm25_24h_co2': '4.8', 'pm10_co2': '4.1', 'pm10_24h_co2': '5.2', 'co2': '476', 'co2_24h': '482', 'lightning_num': '158', 'lightning': '1', 'lightning_time': '1651086083', 'wh80batt': '3.28', 'batt1': '0', 'batt2': '0', 'batt3': '0', 'batt4': '0', 'batt5': '0', 'batt6': '0', 'batt7': '0', 'batt8': '0', 'soilbatt1': '1.3', 'soilbatt2': '1.3', 'soilbatt3': '1.3', 'soilbatt4': '1.3', 'soilbatt5': '1.3', 'pm25batt1': '2', 'wh57batt': '3', 'co2_batt': '6', 'wh90batt': '3.22', 'freq': '868M', 'model': 'GW2000A'}
DEBUG:ecowitt2mqtt:Wind Chill Temperature is only defined for temperatures at or below 50 F and wind speeds above 3 mph. (temperature: 57.6, wind speed: 4.92)
DEBUG:ecowitt2mqtt:Publishing according to Home Assistant MQTT Discovery standard
DEBUG:ecowitt2mqtt:Sending CONNECT (u1, p1, wr0, wq0, wf0, c1, k60) client_id=b''
DEBUG:ecowitt2mqtt:Received CONNACK (0, 0)
INFO:ecowitt2mqtt:No entity description found for key: runtime
INFO:ecowitt2mqtt:No entity description found for key: runtime
DEBUG:ecowitt2mqtt:Sending DISCONNECT
ERROR:aiohttp.server:Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
    resp = await request_handler(request)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/usr/src/ecowitt2mqtt/mqtt.py", line 114, in async_publish_payload
    await _async_publish_to_hass_discovery(client, data, discovery_manager)
  File "/usr/src/ecowitt2mqtt/mqtt.py", line 36, in _async_publish_to_hass_discovery
    _generate_payload("online"),
  File "/usr/src/ecowitt2mqtt/mqtt.py", line 77, in _generate_payload
    return value.encode("utf-8")
UnboundLocalError: local variable 'value' referenced before assignment
INFO:aiohttp.access:10.1.10.160 [27/Apr/2022:21:46:16 +0000] "POST /data/report HTTP/1.1" 500 244 "-" "-"
/usr/local/lib/python3.9/site-packages/aiohttp/web_protocol.py:449: RuntimeWarning: coroutine 'Client.publish' was never awaited
  reset = await self.finish_response(request, resp, start_time)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```
